client/src/selectionRange.proposed.ts

```diff
@@ -84,84 +84,89 @@ export class SelectionRangeFeature extends TextDocumentFeature<TextDocumentRegis
 		});
 	}
 
-	private asSelectionRanges(selectionRanges: SelectionRange[][] | null): VSelectionRange[][] {
+	private asSelectionRanges(selectionRanges: SelectionRange[] | null): VSelectionRange[] {
 		if (!Array.isArray(selectionRanges)) {
 			return [];
 		}
-		let result:  VSelectionRange[][] = [];
-		for (let ranges of selectionRanges) {
-			if (!Array.isArray(ranges)) {
+		let result: VSelectionRange[] = [];
+		for (let range of selectionRanges) {
+			let vrange = this.asSelectionRange(range);
+			if (vrange == null) {
 				return [];
 			}
-			let inner: VSelectionRange[] = [];
-			result.push(inner);
-			for (let range of ranges) {
-				inner.push(new VSelectionRange(
-					this._client.protocol2CodeConverter.asRange(range.range),
-					this.asSelectionRangeKind(range.kind),
-				));
-			}
+			result.push(vrange);
 		}
 		return result;
 	}
 
-	private asSelectionRangeKind(kind?: string): VSelectionRangeKind {
-		switch (kind) {
-			case SelectionRangeKind.Empty:
-				return VSelectionRangeKind.Empty;
-			case SelectionRangeKind.Statement:
-				return VSelectionRangeKind.Statement;
-			case SelectionRangeKind.Declaration:
-				return VSelectionRangeKind.Declaration;
-			default:
-				return VSelectionRangeKind.Empty
+	private asSelectionRange(selectionRange: SelectionRange): VSelectionRange | undefined {
+		let range = this._client.protocol2CodeConverter.asRange(selectionRange.range);
+		let parent = undefined;
+		if (selectionRange.parent) {
+			parent = this.asSelectionRange(selectionRange.parent);
+			if (!parent || !(parent.range.contains(range) && !parent.range.isEqual(range))) {
+				return undefined
```

Comment:
What is the proper way to report invalid data? eg, when parent is smaller than a child?

Reply:
@jrieken any recommendation?

Reply:
Yeah, child must be contained by parents

Reply:
Yep, and this code checks this conditions. The question is, what to do if the server implementation returns wrong data? 

The behavior should be such that:

* server implementer ideally should see a loud error and fix the problem, instead of spending time with debugger to figure out where exactly the info is dropped
* editor user should ideally get *some* message (such that they can open an issue for the corresponding extension), but it should not be distracting, and it should not be repeating. 


Is there some function like `show_error_dialog_if_in_dev_mode_otherwise_log_a_warning`?